### PR TITLE
Benchsolve

### PR DESCRIPTION
**Pull request by jgdumas**
Benchmarking sequential dense system solving (modular or rational)

### CODE REVIEW

File: benchmarks/benchmark-dense-solve.C [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+            chrono.stop();
+
+#ifdef _BENCHMARKS_DEBUG_
+            std::clog << "(DenseElimination) Solution is [";
+            for(const auto& it: X)
+                F.write(std::clog, it) << ' ';
+            std::clog << ']' << std::endl;
+#endif
+            timebits[iter].second=Givaro::logtwo(GIVMAX(X.front(), 1));
+        } else {
```
Comment (ClementPernet) [inline benchmarks/benchmark-dense-solve.C]: Avoid duplicating the same code for the 2 instances: Modular<double> and ZRing<Integer>. Maybe write the same code in a `run_with_field` method templated by the Field type and have `main` call this function twice in a if then else block.

File: benchmarks/benchmark-dense-solve.C [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            typedef Givaro::Modular<double> Field;
+            Field F(q);
+            Field::RandIter G(F);
+	
+#ifdef _BENCHMARKS_DEBUG_
+            std::clog << "Setting A ... " << std::endl;
+#endif
+            chrono.start();
+            DenseMatrix<Field> A(F,n,n);
+            PAR_BLOCK { FFLAS::pfrand(F,G, n,n,A.getPointer(),n); }
```
[the remark this answers is not in the document]
Comment (jgdumas) [reply benchmarks/benchmark-dense-solve.C]: RandomMatrixWithRank etc. are not parallel

File: benchmarks/benchmark-dense-solve.C (hunk 1 above)
Comment (jgdumas) [reply benchmarks/benchmark-dense-solve.C]: The code is not the same, randiter and solve have different API

Comment (Breush) [reply benchmarks/benchmark-dense-solve.C]: Using rational interface (X = BlasVector<Rational>) should do it, or do I miss another difference?

File: benchmarks/benchmark-dense-solve.C [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            chrono.stop();
+#ifdef _BENCHMARKS_DEBUG_
+            std::clog << "... A is " << A.rowdim() << " by " << A.coldim() << ", " << chrono << std::endl;
+            if (A.rowdim() <= 20 && A.coldim() <= 20) A.write(std::clog <<"A:=",Tag::FileFormat::Maple) << ';' << std::endl;
+#endif            
+            DenseVector<Field> X(F, A.coldim()),B(F, A.rowdim());
+            for(auto it=B.begin(); it != B.end(); ++it)
```
[the remark this answers is not in the document]
Comment (jgdumas) [reply benchmarks/benchmark-dense-solve.C]: done.

File: benchmarks/benchmark-dense-solve.C (hunk 1 above)
Comment (jgdumas) [reply benchmarks/benchmark-dense-solve.C]: Hum then X elements do not have the same interface either (I want the bitsize of the denominator)

File: benchmarks/benchmark-dense-solve.C (hunk 2 above)
[the remark this answers is not in the document]
Comment (ClementPernet) [reply benchmarks/benchmark-dense-solve.C]: Ok, I concede to call frand with a try catch block to catch the singular/failed/inconsistent case.

File: benchmarks/benchmark-dense-solve.C (hunk 1 above)
Comment (jgdumas) [reply benchmarks/benchmark-dense-solve.C]: OK, I concede to factorize the code with template specializations

**Comment by ClementPernet** [comment]
LGTM

**Pull request merged**